[Blog] Swift Configuration 1.0 released (#1264)
apple/swift-org-website: _posts/2025-12-11-swift-configuration-1.0-released.md

- **Build and share custom configuration providers** using a public ConfigProvider protocol that anyone can implement and share. This allows domain experts to create integrations with external systems like secret stores and feature flagging services.

Swift Configuration excels in the Swift server ecosystem, where configuration is often read from multiple systems and tools. The library is equally useful in command-line tools, GUI applications, and libraries wherever flexible configuration management is needed.

For a step-by-step evolution of an example service, from hardcoded values all the way to a flexible provider hierarchy, check out the [video of my talk](https://www.youtube.com/watch?v=I3lYW6OEyIs) from the [ServerSide.swift conference](https://www.serversideswift.info/) in London.

## Getting started

After adding a package dependency to your project, reading configuration values requires just a couple of lines of code. For example:

```swift
import Configuration

let config = ConfigReader(provider: EnvironmentVariablesProvider())

> let timeout = config.bool(forKey: "http.timeout", default: 60)

Author: ```suggestion
let timeout = config.int(forKey: "http.timeout", default: 60)
```

> let timeout = config.bool(forKey: "logging.verbose", default: false)

Reviewer: Let's update the key and value be consistent with later mention in the same post.

---
layout: new-layouts/post
published: true
date: 2025-12-11 10:00:00
title: "Swift Configuration 1.0 released"
author: [honzadvorsky]
category: "Developer Tools"
---

> Every application has configuration: in environment variables, configuration files, values from remote services, command-line flags, or repositories for stored secrets like API keys. But until now, Swift developers have had to wire up each source individually, with parsing logic scattered throughout codebases and application code that is tightly coupled to specific configuration providers. 

Reviewer 2: ```suggestion
Every application has configuration: in environment variables, configuration files, values from remote services, command-line flags, or repositories for stored secrets like API keys. But until now, Swift developers have had to wire up each source individually, with scattered parsing logic and application code that is tightly coupled to specific configuration providers. 
```

What do you think of that second sentence revision? The "scattered throughout codebases and application code..." sounded awkward and didn't parse easily for me when I read through it.

Reviewer 3: I like anything that adds clarity, but I found it a bit harder to separate out which bits the "scattered" applied to in your revision. Is it "scattered _parsing logic and application code_", or "scattered _parsing logic_... and application code that..."? The latter is intended. Perhaps we can find a third approach?

[**Swift Configuration**](https://github.com/apple/swift-configuration) brings a unified, type-safe approach to this problem for Swift applications and libraries. What makes this compelling isn’t just that it reads configuration files: plenty of libraries do that. It’s the clean abstraction that it introduces between _how_ your code accesses configuration and _where_ that configuration comes from. This separation unlocks something powerful: libraries can now accept configuration without dictating the source, making them genuinely composable across different deployment environments.

> With the release of Swift Configuration 1.0, the library is production-ready to serve as the common API for reading configuration across the Swift ecosystem. Since the [initial release announcement](https://forums.swift.org/t/introducing-swift-configuration/82368) in October 2025 over **40 pull requests** have been merged, and its API stability provides a foundation to unlock community integrations..

Reviewer 4: I agree with Joe.

As an iOS developer when I read the initial announcement I thought: ‘oh does this work with UserDefaults?’

And as someone who maintains a command-line tool that makes heavy use of Argument Parser, but which also grabs a few environment variables for configuration, my second thought was ‘oh does this work with Argument Parser?’.

(The command-line tool makes heavy use of Argument Parser for subcommands, help, etc., so moving away from it is not feasible.)

So, I think it’s important not to overpromise the capabilities of 1.0, and possibly even mention these current limitations.

> With the release of Swift Configuration 1.0, the library is production-ready to serve as the common API for reading configuration across the Swift ecosystem. Since the [initial release announcement](https://forums.swift.org/t/introducing-swift-configuration/82368) in October 2025 over **40 pull requests** have been merged, and its API stability provides a foundation to unlock community integrations..

Reviewer 3: I also wondered about dialing this back at first, but then I spotted [this](https://forums.swift.org/t/introducing-swift-configuration/82368/10), which implies that iOS apps are indeed a target. But agree that UserDefaults support is something we should add.

@dempseyatgithub -- I didn't quite understand the CLI point? Are you suggesting that this is a poor fit for CLI apps?

- **Quickly get started with a few lines of code** using simple built-in providers for environment variables, command-line arguments, JSON and YAML files. Later, when your configuration needs require a more sophisticated provider, swap it in easily, without refactoring your existing code.
- **Build and share custom configuration providers** using a public ConfigProvider protocol that anyone can implement and share. This allows domain experts to build and own integrations with external systems like secret stores and feature flagging services.

Swift Configuration excels in the Swift server ecosystem, where configuration is often read from multiple systems and tools. The library is equally useful in command-line tools, GUI applications, and libraries wherever flexible configuration management is needed.

For a step-by-step tour of an example service, from hardcoded values all the way to a flexible provider hierarchy, check out the [video of my talk](https://www.youtube.com/watch?v=I3lYW6OEyIs) from the [ServerSide.swift conference](https://www.serversideswift.info/) in London.

## Getting started

> After adding a package dependency to your project, reading configuration just requires a couple of lines of code. For example:

Reviewer 3: Best of both worlds? Maybe "configuration values requires just a couple of lines of code"?

```swift
import Configuration

let config = ConfigReader(provider: EnvironmentVariablesProvider())
let timeout = config.bool(forKey: "logging.verbose", default: false)
```

> But Swift Configuration's core strength is its ability to combine _multiple_ configuration providers into a clear, predictable hierarchy, allowing you to establish sensible defaults while providing clean override mechanisms for different deployment scenarios.

Reviewer: ```suggestion
However, Swift Configuration's core strength is its ability to combine _multiple_ configuration providers into a clear, predictable hierarchy, allowing you to establish sensible defaults while providing clean override mechanisms for different deployment scenarios.
```

> But Swift Configuration's core strength is its ability to combine _multiple_ configuration providers into a clear, predictable hierarchy, allowing you to establish sensible defaults while providing clean override mechanisms for different deployment scenarios.

Reviewer 2: ```suggestion
This above example shows a single configuration provider. Swift Configuration's core strength is its ability to combine _multiple_ configuration providers into a clear, predictable hierarchy, allowing you to establish sensible defaults while providing clean override mechanisms for different deployment scenarios.
```

reframing a smidge to remove the "But..." which I immediately took with a slightly negative connotation that didn't match what was being presented.

Reviewer 3: "However", instead of "but", perhaps? It's intended that the previous example focuses on ease of use and the upcoming example shows the hidden depths...

For example, if you have default configuration in JSON:

```json
{
  "http": {
    "timeout": 30
  }
}
```

And want to be able to provide an override using an environment variable:

```bash
# Environment variables:
HTTP_TIMEOUT=15
```

> 

Reviewer 2: In reading this, you set up and "If..." and "And ...", so I was expecting a "then..." statement of some kind here. The examples are good, but it might flow more easily if you highlight that wanting to provide JSON data and ENV vars for inputs, then introduce the example that shows the multiple readers working together.

Author: Good thought, tried to reword in 884eace4

    * [Peekaboo](https://github.com/steipete/Peekaboo)
    * [swiftodon](https://github.com/JonPulfer/swiftodon)
* [**By implementing custom providers**](https://swiftpackageindex.com/apple/swift-configuration/main/documentation/configuration/configprovider): Extend the ecosystem with new formats or external sources by implementing a [`ConfigProvider`](https://swiftpackageindex.com/apple/swift-configuration/main/documentation/configuration/configprovider). Examples of experimental providers:
    * [Swift Configuration TOML](https://github.com/finnvoor/swift-configuration-toml)
    * [Vault Courier](https://github.com/vault-courier/vault-courier)
    * [Swift Configuration AWS](https://github.com/SongShift/swift-configuration-aws)

Contributions and new integrations are welcome.

## Next steps

With a stable foundation in place, libraries and applications can begin finalizing their own integrations and releasing API-stable versions built on Swift Configuration.

> We encourage you to continue sharing real-world experience across different deployment models and configuration patterns. Your feedback will continue to help guide future refinements of Swift Configuration.

Author: Merged in 3a6f2381
